src/librustc_lint/unused.rs

```diff
@@ -464,3 +464,61 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedAllocation {
         }
     }
 }
+
+declare_lint! {
+    pub(super) UNUSED_LABEL,
+    Warn,
+    "warns on unused labels"
+}
+
+#[derive(Clone)]
+pub struct UnusedLabel(pub Vec<ast::Label>);
+
+impl UnusedLabel {
+    pub fn new() -> Self {
+        UnusedLabel(vec![])
+    }
+}
+
+impl LintPass for UnusedLabel {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNUSED_LABEL)
+    }
+}
+
+impl EarlyLintPass for UnusedLabel {
+    fn check_expr(&mut self, _: &EarlyContext, expr: &ast::Expr) {
+        match expr.node {
+            ast::ExprKind::While(_, _, Some(label))
+            | ast::ExprKind::WhileLet(_, _, _, Some(label))
+            | ast::ExprKind::ForLoop(_, _, _, Some(label))
+            | ast::ExprKind::Loop(_, Some(label)) => {
+                self.0.push(label);
+            }
+            ast::ExprKind::Break(Some(label), _) | ast::ExprKind::Continue(Some(label)) => {
+                if let Some(index) = self.0.iter().rposition(|&l| l.ident == label.ident) {
+                    self.0.remove(index);
```

Comment:
Add another comment along the lines of "Remove labels as they're used. If they're used more than once they will already have been removed, as expected".